format-delimited/src/main/java/io/cdap/plugin/format/delimited/input/DelimitedConfig.java

```diff
@@ -122,47 +179,101 @@ public Schema getDefaultSchema(@Nullable FormatContext context) throws IOExcepti
 
     Job job = JobUtils.createInstance();
     Configuration conf = job.getConfiguration();
-    // set entries here, before FileSystem is used
+
     for (Map.Entry<String, String> entry : getFileSystemProperties().entrySet()) {
       conf.set(entry.getKey(), entry.getValue());
     }
-    FSDataInputStream input = null;
-    BufferedReader bufferedReader = null;
+    final Path file = getFilePathForSchemaGeneration(path,
+                                                     format.equals(DELIMITER) ? null : format, conf);
+    final FileSystem fileSystem = FileSystem.get(file.toUri(), conf);
+
     String line = null;
-    try {
-      final Path file = getFilePathForSchemaGeneration(path,
-                                                       format.equals("delimited") ? null : format, conf);
-      final FileSystem fileSystem = FileSystem.get(file.toUri(), conf);
-      input = fileSystem.open(file);
-      bufferedReader = new BufferedReader(new InputStreamReader(input));
-      line = bufferedReader.readLine();
-      if (line == null) {
-        return null;
+    String[] columnNames = null;
+    String[] columnValues = null;
+
+    try (FSDataInputStream input = fileSystem.open(file);
+         BufferedReader bufferedReader = new BufferedReader(new InputStreamReader(input))) {
+      for (int rowIndex = 0; rowIndex < sampleSize && (line = bufferedReader.readLine()) != null; rowIndex++) {
+        columnValues = line.split(getDefaultDelimiter(), -1);
+
+        if (rowIndex == 0) {
+          columnNames = setColumnNames(override, line, columnNames, columnValues);
+          continue;
+        }
+        runInferValueDataType(override, dataTypeInvestigator, columnNames, columnValues);
       }
-    } finally {
-      if (bufferedReader != null) {
-        bufferedReader.close();
+    } catch (IOException e) {
+      throw new RuntimeException(e);
+    }
+
+    runInferColumnDataType(override, dataTypeInvestigator, fields, columnNames);
+    return Schema.recordOf("text", fields);
+  }
+
+  /**
+   * Set column names. If the header is skipped, column names are generated automatically as body_1, body_2,..., body_k.
+   * If the header is kept, the first line of the dataset is taken to specify the column names.
+   * @param override
+   * @param line
+   * @param columnNames
+   * @param columnValues
+   */
+  private String[] setColumnNames(HashMap<String, Schema> override, String line, String[] columnNames,
+                                  String[] columnValues) {
+    if (!getSkipHeader()) {
+      columnNames = line.split(getDefaultDelimiter());
```

Comment:
I formatted the code now, and It should be clearer.

To explain the reason behind, consider that you have a CSV dataset ready to infer the schema from.

1. If `keepHeader=True` we split the first row with `limit=0`, and every other record with `limit=-1`. This means:
       - The raw header:`"name,surname,age,address"`  - resolves to - `name`, `surname`,`age`,`address` fields.
       - The value records:`"john,doe,,"` - resolves to - `john`, `doe`, `(empty)`, `(emtpy)`. Because of `limit=-1`, the empty  values at the end of the line are considered when detecting the column data type.
      
2. If `keepHeader=False`, every record is split with `limit=-1`. This means:
       -  If the first record is `"john,doe,,"`, the output schema will have `body_1`, `body_2`, `body_3`, `body_4` fields.

Reply:
I see, thanks for the explanation!